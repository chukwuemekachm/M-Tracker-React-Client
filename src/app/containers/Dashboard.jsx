import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import DefaultNavBar from '../components/NavBar';
import userRequestActions from '../actions/userRequestActions';
import UserRequest from '../components/UserRequest';
import '../assets/css/dashboard.css';
import SideNav from '../components/SideNav';
import CreateRequestForm from '../components/CreateRequest';
import filterRequests, { navigateFilter } from '../actions/filterRequestsAction';
import getAllRequests from '../actions/adminRequestActions';

/**
 * @description Dashboard component
 *
 * @class
 *
 * @extends React.Component
 */
export class Dashboard extends Component {
  /**
   * @constructor
   *
   * @param {object} props The injected props
   */
  constructor(props) {
    super(props);
    this.state = {};
  }

  /**
   * @description Evaluates the user's authentication
   * Calls the getRequest method to fetch the user's request
   *
   * @function
   */
  componentDidMount() {
    const {
      authenticated, history, getRequests, filter, admin, getAdminRequests,
    } = this.props;
    if (!authenticated) {
      return history.push('/login');
    }
    if (admin) {
      return getAdminRequests().then((response) => {
        if (response.code === 401) return history.push('/login');
        const { search } = history.location;
        return filter(search.substring(1));
      });
    }
    return getRequests().then((response) => {
      if (response.code === 401) return history.push('/login');
      const { search } = history.location;
      return filter(search.substring(1));
    });
  }

  /**
   * @description Renders the component on a dom node
   *
   * @function
   *
   * @returns {object} The components to render
   */
  render() {
    const {
      requests, history, createRequest, filteredRequests,
    } = this.props;
    return (
      <div>
        <div className="dashboard-body">
          <DefaultNavBar />
          <div className="container mt-5">
            <div className="row">
              <div className="col-md-3">
                <SideNav requests={requests} handleNavigation={navigateFilter} />
              </div>
              <div className="col-md-8">
                {
                  filteredRequests[0]
                    ? filteredRequests.map(request => (
                      <Link to={`/${request.id}`} key={request.id}>
                        <UserRequest key={request.id} {...request} />
                      </Link>
                    ))
                    : (
                      <div>
                        {
                          requests[0] && !filteredRequests[0]
                            ? (
                              <p className="text-center">
                                No requests exist for this category
                              </p>
                            )
                            : (
                              <p className="text-center">
                                You have no requests please click the
                                {' '}
                                create request button to make a new request.
                              </p>
                            )
                        }
                      </div>
                    )
                }
              </div>
            </div>
            <div className="col-md-1" />
          </div>
        </div>
        <CreateRequestForm history={history} create={createRequest} />
      </div>
    );
  }
}

Dashboard.propTypes = {
  requests: PropTypes.arrayOf(PropTypes.object).isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
  getRequests: PropTypes.func.isRequired,
  createRequest: PropTypes.func.isRequired,
  filter: PropTypes.func.isRequired,
  filteredRequests: PropTypes.arrayOf(PropTypes.object).isRequired,
  getAdminRequests: PropTypes.func.isRequired,
  admin: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state, ownprops) => ({
  requests: state.requests.allRequests,
  filteredRequests: state.requests.filteredRequests,
  authenticated: state.auth.authenticated,
  history: ownprops.history,
  admin: state.auth.user.admin,
});

export const mapDispatchToProps = dispatch => ({
  getAdminRequests: () => dispatch(getAllRequests()),
  getRequests: () => dispatch(userRequestActions.getAllAsync()),
  createRequest: payload => dispatch(userRequestActions.createRequest(payload)),
  filter: payload => dispatch(filterRequests(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
