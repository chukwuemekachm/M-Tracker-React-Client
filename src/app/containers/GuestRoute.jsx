import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

/**
 * @description Has the same API as the Route component
 * It verifies if the user is authenticated or not
 * If the user is logged in it redirects to the dashboard
 * Else it renders the component
 *
 * @param {object} component The component to be rendered
 * @param {boolean} authenticated The authentication state of the user
 * @param {array} rest Other props injected to the component
 *
 * @returns {object}
 * */
export const GuestRoute = ({ component: Component, authenticated, ...rest }) => (
  <Route
    {...rest}
    render={props => (
      !authenticated
        ? <Component {...props} />
        : <Redirect to="/dashboard" />
    )}
  />
);

GuestRoute.propTypes = {
  component: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  authenticated: state.auth.authenticated,
});

export default connect(mapStateToProps, null)(GuestRoute);
