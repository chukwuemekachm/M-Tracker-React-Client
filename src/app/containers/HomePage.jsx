import React from 'react';
import { Link } from 'react-router-dom';
import DefaultNavBar from '../components/NavBar';
import Icon from '../components/Icon';
import Footer from '../components/Footer';
import '../assets/css/home.css';

/**
 * @description The Home page component
 *
 * @returns {object}
 */
const HomePage = () => (
  <div>
    <DefaultNavBar />
    <div className="jumbotron">
      <div className="banner">
        <div className="w3-display-middle">
          <h3 className="text-center">
            Facilities don
            {'\''}
            t need to go bad any longer,
            {' '}
            request a repair or maintenance schedule from us today.
          </h3>
          <br />
          <div className="row">
            <div className="col-md-3" />
            <div className="col-md-6">
              <Link to="/signup">
                <button type="button" className="btn btn-success btn-lg btn-block">
                  Get Started
                </button>
              </Link>
            </div>
            <div className="col-md-3" />
          </div>
        </div>
      </div>
    </div>
    <div className="container">
      <br />
      <h3 className="display-6 text-center">
        About Maintenance Tracker
      </h3>
      <br />
      <div className="container">
        <p className="text-center">
          Maintenance Tracker is an platform designed, to provide users with the oppurtunity of
          {' '}
          reaching out to operations or repairs department regarding repair or maintenance
          {' '}
          requests. It also enables users to monitor the status of their requests.
        </p>
      </div>
    </div>
    <br />
    <br />
    <div className="container">
      <h3 className="display-6 text-center">
        How it Works
      </h3>
      <br />
      <br />
      <div className="row w3-display-center">
        <Icon
          title="Authenticate"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_login_jdch.svg"
        />
        <Icon
          title="Log a request"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265099/undraw_add_file2_gvbb.svg"
        />
        <Icon
          title="Request gets approved"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_hiring_cyhs.svg"
        />
        <Icon
          title="Request gets resolved"
          icon="https://res.cloudinary.com/dvcaeuvee/image/upload/v1538265100/undraw_confirmation_2uy0.svg"
        />
      </div>
    </div>
    <Footer />
  </div>
);

export default HomePage;
