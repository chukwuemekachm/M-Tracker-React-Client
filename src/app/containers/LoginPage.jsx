import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import NavBar from '../components/NavBar';
import LoginForm from '../components/LoginForm';
import Alert from '../components/Alert';
import authActions from '../actions/authActions';
import loadingSvg from '../assets/svg/loading.svg';

export class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const {
      error, message, loading, history, userLogin,
    } = this.props;
    return (
      <div>
        <NavBar />
        <div className="ch-row">
          <h2 className="ch-about">
            Login to access your Maintenance Tracker Account
          </h2>
        </div>
        <div className="ch-row">
          <div className="ch-col-3">
            {
              error
                ? (
                  <Alert
                    message={message}
                    backgroundColor="#E74C3C"
                  />
                )
                : ''
            }
            {
              loading
                ? <img src={loadingSvg} alt="loading" />
                : ''
            }
          </div>
        </div>
        <div className="ch-row">
          <div className="ch-col-3 ch-panel ch-signup">
            <LoginForm
              userLoginRequest={userLogin}
              history={history}
            />
            <p className="ch-right">
              {'Don\'t '}
              have an account?
              {' '}
              <Link to="/signup">
                Signup...
              </Link>
            </p>
          </div>
        </div>
      </div>
    );
  }
}

LoginPage.propTypes = {
  message: PropTypes.string.isRequired,
  error: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  userLogin: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  error: state.common.error,
  message: state.common.message,
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  userLogin: (user, history, route) => (authActions.authAsync(user, history, route)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
