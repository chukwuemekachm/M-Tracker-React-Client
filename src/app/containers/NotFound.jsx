import React from 'react';
import DefaultNavBar from '../components/NavBar';

/**
 * @description The Not found component
 *
 * @returns {object}
 */
const NotFound = () => (
  <div className="ch-aside">
    <DefaultNavBar />
    <div
      className="ch-row"
      style={{
        backgroundColor: 'rgba(0, 0, 0, 0.7)', color: 'white', width: '100%', minHeight: '100%', position: 'fixed',
      }}
    >
      <div className="ch-col-6">
        <h1 style={{ fontSize: '8em' }}>
          404
        </h1>
        <p>
          Oops, the page you are looking for was not found.
        </p>
      </div>
    </div>
  </div>
);

export default NotFound;
