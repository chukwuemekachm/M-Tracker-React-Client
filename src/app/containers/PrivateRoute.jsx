import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

/**
 * @description Has the same API as the Route component
 * It verifies if the user is authenticated or not
 *
 * @param {object} component The component to be rendered
 * @param {boolean} authenticated The authentication state of the user
 * @param {array} rest Other props injected to the component
 *
 * @returns {object}
 * */
export const PrivateRoute = ({ component: Component, authenticated, ...rest }) => (
  <Route
    {...rest}
    render={props => (
      authenticated
        ? <Component {...props} />
        : <Redirect to="/login" />
    )}
  />
);

PrivateRoute.propTypes = {
  component: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  authenticated: state.auth.authenticated,
});

export default connect(mapStateToProps, null)(PrivateRoute);
