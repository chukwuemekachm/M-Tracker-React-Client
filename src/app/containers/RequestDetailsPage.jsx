import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import SideNav from '../components/SideNav';
import NavBar from '../components/NavBar';
import { userError } from '../actions/commonActions';
import RequestDetails from '../components/RequestDetails';

export class RequestDetailsPage extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount() {
    const { isAuthenticated, history } = this.props;
    if (!isAuthenticated) {
      history.push('/login');
    }
  }

  /* eslint-disable class-methods-use-this */
  formatButtons(status) {
    let buttons;
    switch (status) {
      case 'approved':
        buttons = '<i class="icon ion-md-arrow-back"></i>';
        break;
      case 'disapproved':
        buttons = `<i class="icon ion-md-arrow-back"></i>
      <i class="icon ion-md-trash"></i>`;
        break;
      case 'resolved':
        buttons = `<i class="icon ion-md-arrow-back"></i>
      <i class="icon ion-md-trash"></i>`;
        break;
      default:
        buttons = `<i class="icon ion-md-arrow-back"></i>
      <i class="icon ion-md-create"></i>`;
        break;
    }
    return { __html: buttons };
  }

  render() {
    const { request } = this.props;
    return (
      <div>
        <NavBar />
        <SideNav />
        <div className="ch-row">
          <div className="ch-col-2" />
          <div className="ch-col-9" id="user-display">
            <h2 className="ch-display">
              <i className="icon ion-md-speedometer" />
              {' '}
              Dashboard
            </h2>
            <RequestDetails {...request} actions={this.formatButtons(request.status)} />
          </div>
        </div>
      </div>
    );
  }
}

RequestDetailsPage.propTypes = {
  request: PropTypes.shape({}).isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state, ownprops) => ({
  request: state.requests.currentRequest,
  isAuthenticated: state.auth.authenticated,
  history: ownprops.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  userErrorHandler: message => (userError(message)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(RequestDetailsPage);
