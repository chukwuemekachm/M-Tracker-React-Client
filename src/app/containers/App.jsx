import React from 'react';
import HomePage from './HomePage';
import '../assets/css/style.css';

const App = () => (
  <div>
    <HomePage />
  </div>
);

export default App;
