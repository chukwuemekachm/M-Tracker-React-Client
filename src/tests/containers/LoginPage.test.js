import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../app/containers/LoginPage';

const userLogin = jest.fn();
const history = {};

describe('LoginPage container', () => {
  it('should render without errors when props are falsy', () => {
    const wrapper = shallow(<LoginPage
      loading={false}
      error={false}
      message="test"
      userLogin={userLogin}
      history={history}
    />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without errors when props are truthy', () => {
    const wrapper = shallow(<LoginPage
      loading
      error
      message="test"

      userLogin={userLogin}
      history={history}
    />);
    expect(wrapper).toBeTruthy();
  });
});
