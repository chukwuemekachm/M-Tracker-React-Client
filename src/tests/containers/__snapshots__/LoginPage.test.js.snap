// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage container should render without errors when props are falsy 1`] = `
<div>
  <Connect(NavBar) />
  <div
    className="ch-row"
  >
    <h2
      className="ch-about"
    >
      Login to access your Maintenance Tracker Account
    </h2>
  </div>
  <div
    className="ch-row"
  >
    <div
      className="ch-col-3"
    />
  </div>
  <div
    className="ch-row"
  >
    <div
      className="ch-col-3 ch-panel ch-signup"
    >
      <LoginForm
        history={Object {}}
        userLoginRequest={[MockFunction]}
      />
      <p
        className="ch-right"
      >
        Don't 
        have an account?
         
        <Link
          replace={false}
          to="/signup"
        >
          Signup...
        </Link>
      </p>
    </div>
  </div>
</div>
`;
